crates/ruff/src/rules/ruff/rules/invalid_index_type.rs

```diff
@@ -0,0 +1,176 @@
+/// other types will result in a `TypeError`.
+///
+/// ## Example
+/// ```python
+/// var = [1, 2, 3]["x"]
+/// ```
+///
+/// Use instead:
+/// ```python
+/// var = [1, 2, 3][0]
+/// ```
+#[violation]
+pub struct InvalidIndexType {
+    var_type: String,
+    idx_type: String,
+    slice_bound: bool,
+}
+
+impl Violation for InvalidIndexType {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let InvalidIndexType {
+            var_type, idx_type, ..
+        } = self;
+        if self.slice_bound {
+            format!("Slice in indexed access to type `{var_type}` uses type `{idx_type}` as bound instead of an integer.")
+        } else {
+            format!(
+                "Indexed access to type `{var_type}` uses type `{idx_type}` instead of an integer or slice."
+            )
+        }
+    }
+}
+
+/// RUF015
+/// Expects components of a `Subscript` expression
+pub(crate) fn invalid_index_type<'a>(checker: &mut Checker, value: &'a Expr, slice: &'a Expr) {
+    // If the value being indexed is a list, tuple, string, f-string, bytes, or comprehension
+    if matches!(
+        value,
+        Expr::List(_)
+            | Expr::ListComp(_)
+            | Expr::Tuple(_)
+            | Expr::JoinedStr(_)
+            | Expr::Constant(ExprConstant {
+                value: Constant::Str(_) | Constant::Bytes(_),
+                ..
+            })
+    ) {
+        // Then check the contents of the index
+        match slice {
+            // If the index is a const, only allow integers
+            Expr::Constant(ExprConstant {
+                value: index_value, ..
+            }) => {
+                if !index_value.is_int() {
+                    checker.diagnostics.push(Diagnostic::new(
+                        InvalidIndexType {
+                            var_type: expression_type_name(value)
+                                .expect("Failed to cast parent expression to type name"),
+                            idx_type: constant_type_name(index_value),
+                            slice_bound: false,
+                        },
+                        slice.range(),
+                    ));
+                }
+            }
+            // If the index is a slice, check for integer or null bounds
+            Expr::Slice(ExprSlice { lower, upper, .. }) => {
+                for slice_bound in [lower, upper].into_iter().flatten() {
+                    if let Expr::Constant(ExprConstant {
+                        value: index_value, ..
+                    }) = slice_bound.as_ref()
+                    {
+                        if !(index_value.is_int() || index_value.is_none()) {
+                            checker.diagnostics.push(Diagnostic::new(
+                                InvalidIndexType {
+                                    var_type: expression_type_name(value)
+                                        .expect("Failed to cast parent expression to type name"),
```

Comment:
For example, you could introduce a new type like this:

```rust
#[derive(Debug)]
enum IndexType<'a> {
    Constant(&'a Constant),
    JoinedStr,
    List,
    ListComp,
    DictComp,
    Set,
    Dict,
    Tuple,
}

impl fmt::Display for IndexType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Self::Constant(constant) => f.write_str(&constant_type_name(constant)),
            Self::JoinedStr => f.write_str("str"),
            Self::List => f.write_str("list"),
            Self::ListComp => f.write_str("list comprehension"),
            Self::DictComp => f.write_str("dict comprehension"),
            Self::Set => f.write_str("set"),
            Self::Dict => f.write_str("dict"),
            Self::Tuple => f.write_str("tuple"),
        }
    }
}
```

(Or replace the constant variant with `ConstantStr` and `ConstantBytes`.)

Then, with a method like:

```rust
impl<'a> IndexType {
    fn try_from(expr: &'a Expr) -> Option<Self<'a>> {
        match expr {
            Expr::Constant(ExprConstant { value, .. }) => Some(Self::Constant(value)),
            Expr::JoinedStr(_) => Some(Self::JoinedStr),
            Expr::List(_) => Some(Self::List),
            Expr::ListComp(_) => Some(Self::ListComp),
            Expr::DictComp(_) => Some(Self::DictComp),
            Expr::Set(_) => Some(Self::Set),
            Expr::Dict(_) => Some(Self::Dict),
            Expr::Tuple(_) => Some(Self::Tuple),
            _ => None,
        }
    }
}
```

If you do this, then the first line of this method could be like:

```rust
let Some(value) = IndexType::try_from(value) else {
    return None;
};

// Maybe some other conditions depending on where you are in the code?
if !matches!(value, IndexType::List | IndexType::Dict | ...) {
   return None;
}
```

Reply:
(There might be some intricacies in this rule that might this pattern not as trivial as I've painted it above, but I'm more trying to demonstrate how I'd think about the logic and responsibilities. How can we use types to make `.expect` and `.unwrap` calls unnecessary or even impossible?)

Reply:
Makes sense, I can definitely do this if it feels worth the time.

Reply:
I think it's worth a try. @MichaReiser may have a different opinion if you want him to weigh in too :)

Reply:
The `IndexType` may also be something that's valuable for other rules and, thus, would make sense to have in our ast crate. But it would then probably make sense that each variant stores the corresponding expression node (similar to `AnyFunctionDefinition`), implements `Ranged`.

Reply:
@MichaReiser that's not an enumeration of valid index types. That's just an enumeration of types that we can reasonably detect as _invalid_. It's more of a `CheckableLiteralType` which could also be useful elsewhere?

Reply:
Oh whoops. I like charlie's proposal because it's explicit about which types can exist (I don't have to infer it from looking at other functions) but your solution is fine too.